Core/Battlefield: rework stage 1

ccrs opened the pull request:
**Changes proposed:**

First stage, goals:
* Remove current implementation (Battlefield & Wintergrasp) leaving only placeholders.
* Spawn missing "base" entities.
* Streamline `trinity_strings`

I mainly do this so contributors and db editors can help with the tedious blizzlike spawn "search & find" process of the missing WG entities. For a time there wont be any WG available, but its "for a time" :dancer:

ref #20144 #21995

**Target branch(es):** 3.3.5

**Issues addressed:** Closes #  (insert issue tracker number)

**Tests performed:** (Does it build, tested in-game, etc.)

**Known issues and TODO list:** (add/remove lines as needed)

ccrs committed:
Core/Battlefield: remove previous implementation

ccrs committed:
Core/Server: define BattlefieldPackets

ccrs committed:
Core/Battlefield: class placeholder

jackpoz commented:
What's the roadmap with this ?

ccrs commented:
Well what I do prefer is:

- Remove current implementation
- Add placeholders
- Add basic spawning
    - Remove Horde/Alliance dependant ones
    - Temporaly spawn both neutral Alliance and Horde bases
- Update configuration file

Then when the baseline is finaly set, contributors and members can actually help with "sides" or "team dependant" spawning.
   -> Maybe create an issue with the "help wanted" tag

Thats what I see as the "cleanest" solution, open to suggestions ofc! 👍

ccrs commented:
DB spawning will take some time, moving it to a branch gives this "project" a high risk to simply die down... I will also help on the task ofc but, as can be seen, on a rather slow basis

Shauren commented:
Yeaaah I cant agree with just deleting it in 3.3.5 "for future rewrite"

ccrs committed:
Core/Battlefield: BattlefieldMgr class placeholder

ccrs commented:
Sure, like I said its something totally open to dicussion. In the end its a matter of merging into the correct branch.

AByakugans commented:
Hi, thank you very much for doing this. As I said in previous PR I want to help. I am not sure I can code well but at least test all properly... Just tell me and I do it.

And please, dont do the branches. There are still branches like Blink, bg rewrire and some query just laying there for eternity. Dont want this to happen here.

To Shauren. Removing it is no real harm done right now, it is really not working at all.

jackpoz commented:
The grafana branch was merged just fine. If a developer loses interest into it, it will be unfinished either way if it's on a branch, a PR or on master.
Forcing users to work on it because we removed it doesn't sounds like the way to go. If you want it done, advertise it around (irc, forum, status reports on the PR), recruit people in the team like @AByakugans who volunteered. If there will be interest from the community and/or the devs it will be surely completed (without pissing anyone in the meantime by taking away completely something).

ccrs committed:
Core/Battlefield: BattlefieldMgr changes + rebase corrections

ccrs committed:
Core/Battlefield: complete BattlefieldMgr

jackpoz commented:
how's going with the progress :) ?

ccrs commented:
yday was like the first day I had couple hours of "totally free" time. Rough work loads these days hehe
regarding actual pr progress, the next step is to cleanup wintergrasp references outside the battlefield scope
should be easy, right? 💃

ccrs committed:
Scripts/Northrend: zone_wintergrasp first cleanup

ccrs committed:
its not like im switching to another computer and this commit will disappear shortly.... or anything

AByakugans commented:
Anything for us people to test yet and help?

ccrs committed:
hello there

ccrs commented:
@illfated and anyone willing, I'd love some help with this
https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/Northrend/zone_wintergrasp.cpp#L37

convert into DB values and stuff 😄

illfated commented:
Regarding the `#define` texts, 3 of them are already in DB in form of gossip options:
```
#define GOSSIP_HELLO_DEMO1  "Build catapult."
#define GOSSIP_HELLO_DEMO2  "Build demolisher."
#define GOSSIP_HELLO_DEMO3  "Build siege engine."
```
```sql
SELECT * FROM `gossip_menu_option` WHERE `MenuID` IN (9904, 9923);
```
```
MenuID  OptionID  OptionIcon  OptionText                             OptionBroadcastTextID  OptionType  OptionNpcFlag  ActionMenuID  ActionPoiID  BoxCoded  BoxMoney  BoxText  BoxBroadcastTextID  VerifiedBuild
------  --------  ----------  -------------------------------------  ---------------------  ----------  -------------  ------------  -----------  --------  --------  -------  ------------------  -------------
  9904         0           0  I would like to build a catapult.                      31058           1              1             0            0         0         0                            0              0
  9904         1           0  I would like to build a demolisher.                    31055           1              1             0            0         0         0                            0              0
  9904         2           0  I would like to build a siege engine.                  31061           1              1             0            0         0         0                            0              0
  9923         0           0  I'd like to build a catapult.                          31058           1              1             0            0         0         0                            0              0
  9923         1           0  I'd like to build a demolisher.                        31055           1              1             0            0         0         0                            0              0
  9923         2           0  I'd like to build a siege engine.                      31061           1              1             0            0         0         0                            0              0
```
I am not sure why there are 2 sets of the same lines (maybe 1 for horde, 1 for alliance?) because the client should show only the `OptionBroadcastTextID` which is the content of 9923.

You mostly need to replace the `#define` lines with enums 0, 1 and 2 (`OptionID`) and name the enums something like
```cpp
enum Gossip
{
   GOSSIP_OPTION_CATAPULT,
   GOSSIP_OPTION_DEMOLISHER,
   GOSSIP_OPTION_SIEGE_ENGINE
};
```

When it comes to line 4,
```cpp
#define GOSSIP_HELLO_DEMO4  "I cannot build more!"
```
and
```cpp
                else
                    AddGossipItemFor(player, GOSSIP_ICON_CHAT, GOSSIP_HELLO_DEMO4, GOSSIP_SENDER_MAIN, GOSSIP_ACTION_INFO_DEF + 9);
```
it looks completely made up to me. Why would the player need to click on a text bubble saying "I cannot build more!" instead of just closing a plain gossip window?
It is much more likely that the gossip window (`npc_text`) would contain a plain text saying this, followed by the [Goodbye] button at the bottom.
(The text "I cannot build more!" was not found in TDB 335.64 `broadcast_text`, `npc_text` or `gossip_menu_option` .)

If possible, look for a video (or play retail) to get the actual gossip dialogue where you can't build something, to see what it is supposed to look like.

BTW, changing the code to use the 3 enums could be as simple as this (just replace all GOSSIP_HELLO_DEMO 1,2,3 )
```cpp
                if (CanBuild())
                {
                    if (player->HasAura(SPELL_CORPORAL))
                        AddGossipItemFor(player, GOSSIP_ICON_CHAT, GOSSIP_OPTION_CATAPULT, GOSSIP_SENDER_MAIN, GOSSIP_ACTION_INFO_DEF);
                    else if (player->HasAura(SPELL_LIEUTENANT))
                    {
                        AddGossipItemFor(player, GOSSIP_ICON_CHAT, GOSSIP_OPTION_CATAPULT, GOSSIP_SENDER_MAIN, GOSSIP_ACTION_INFO_DEF);
                        AddGossipItemFor(player, GOSSIP_ICON_CHAT, GOSSIP_OPTION_DEMOLISHER, GOSSIP_SENDER_MAIN, GOSSIP_ACTION_INFO_DEF + 1);
                        AddGossipItemFor(player, GOSSIP_ICON_CHAT, GOSSIP_OPTION_SIEGE_ENGINE, GOSSIP_SENDER_MAIN, GOSSIP_ACTION_INFO_DEF + 2);
                    }
                }
```
(Follow-up changes to the "I cannot build more!" case depends on what the actual gossip looks like.)

illfated commented:
Oh. I need to go back and look for my previous work on gossip, because you also need to include `GOSSIP_MENU_ID` = 9923 (maybe even 9904, not sure).

Please also make a note of any diffs between Alliance and Horde in gossips, unless they are handled already.

illfated commented:
Another preliminary set of enums, we need both GOSSIP_OPTION and NPC_TEXT for DB gossip usage in the script.
```cpp
enum Gossip
{
    GOSSIP_OPTION_CATAPULT          0,
    GOSSIP_OPTION_DEMOLISHER        1,
    GOSSIP_OPTION_SIEGE_ENGINE      2,

    NPC_TEXT_HOW_CAN_I_HELP_H       13759,
    NPC_TEXT_SORRY_ONLY_OFFICER_H   13761,
    NPC_TEXT_HOW_CAN_I_HELP_A       13798,
    NPC_TEXT_SORRY_ONLY_OFFICER_A   14172
};
```
This means we need to include a team condition test within the `bool GossipHello()` call, to differentiate between the gossips (unless it can be done via `player->GetGossipTextId(me)` which I haven't tried myself in any of my script changes).

illfated commented:
Updated preliminary gossip script sample, using proper enums for DB retrieval:
```cpp
                if (CanBuild())
                {
                    if (player->HasAura(SPELL_CORPORAL))
                        AddGossipItemFor(player, NPC_TEXT_HOW_CAN_I_HELP_H, GOSSIP_OPTION_CATAPULT, GOSSIP_SENDER_MAIN, GOSSIP_ACTION_INFO_DEF);
                    else if (player->HasAura(SPELL_LIEUTENANT))
                    {
                        AddGossipItemFor(player, NPC_TEXT_HOW_CAN_I_HELP_H, GOSSIP_OPTION_CATAPULT, GOSSIP_SENDER_MAIN, GOSSIP_ACTION_INFO_DEF);
                        AddGossipItemFor(player, NPC_TEXT_HOW_CAN_I_HELP_H, GOSSIP_OPTION_DEMOLISHER, GOSSIP_SENDER_MAIN, GOSSIP_ACTION_INFO_DEF + 1);
                        AddGossipItemFor(player, NPC_TEXT_HOW_CAN_I_HELP_H, GOSSIP_OPTION_SIEGE_ENGINE, GOSSIP_SENDER_MAIN, GOSSIP_ACTION_INFO_DEF + 2);
                    }
                }
```
I am still uncertain how to make a Team_A / Team_H check and decide how to add the gossip content based on player team.

Killyana commented:
Missing gossip added here https://github.com/TrinityCore/TrinityCore/commit/d755ef1e8b303d285b82c37f9470b3080f641dd3
Every thing is on the DB, #define GOSSIP_HELLO_DEMO4 "I cannot build more!"  looks wrong for me the npc mut just show the text: "Sorry, friend. I'm afraid we've used all of the parts we can spare at the moment."

illfated commented:
Thanks for confirming what I mentioned earlier.
I am pretty sure we can scrap those hardcoded gossip options now.

ccrs committed:
npc_wg_demolisher_engineer

ccrs committed:
more zone_wintergrasp scripts

ccrs committed:
world & configurations

illfated commented:
Just to give a little feedback on the SQL part: it seems you are very, **very** specific in your deletions. Are you worried that the SQL could touch something else than those 16 lines you want to remove? Maybe you want to avoid removing stuff for players using custom conditions? My suggestion would be to simplify the query to remove only what already exists in TDB 335.64 already :
```sql
DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` IN (14, 15) AND `SourceGroup` IN (9904, 9923);
```
Just to show you what we currently have got in TDB 335.64 with the result from
```sql
SELECT * FROM `conditions` WHERE `SourceTypeOrReferenceId` IN (14, 15) AND `SourceGroup` IN (9904, 9923);
```
```
SourceTypeOrReferenceId  SourceGroup  SourceEntry  SourceId  ElseGroup  ConditionTypeOrReference  ConditionTarget  ConditionValue1  ConditionValue2  ConditionValue3  NegativeCondition  ErrorType  ErrorTextId  ScriptName  Comment
-----------------------  -----------  -----------  --------  ---------  ------------------------  ---------------  ---------------  ---------------  ---------------  -----------------  ---------  -----------  ----------  -------
                     14         9904        13759         0          0                         1                0            33280                0                0                  0          0            0              (NULL)
                     14         9904        13759         0          1                         1                0            55629                0                0                  0          0            0              (NULL)
                     14         9904        13761         0          0                         1                0            33280                0                0                  1          0            0              (NULL)
                     14         9904        13761         0          0                         1                0            55629                0                0                  1          0            0              (NULL)
                     14         9923        13798         0          0                         1                0            33280                0                0                  0          0            0              (NULL)
                     14         9923        13798         0          1                         1                0            55629                0                0                  0          0            0              (NULL)
                     14         9923        14172         0          0                         1                0            33280                0                0                  1          0            0              (NULL)
                     14         9923        14172         0          0                         1                0            55629                0                0                  1          0            0              (NULL)
                     15         9904            0         0          0                         1                0            33280                0                0                  0          0            0              (NULL)
                     15         9904            0         0          1                         1                0            55629                0                0                  0          0            0              (NULL)
                     15         9904            1         0          0                         1                0            55629                0                0                  0          0            0              (NULL)
                     15         9904            2         0          0                         1                0            55629                0                0                  0          0            0              (NULL)
                     15         9923            0         0          0                         1                0            33280                0                0                  0          0            0              (NULL)
                     15         9923            0         0          1                         1                0            55629                0                0                  0          0            0              (NULL)
                     15         9923            1         0          0                         1                0            55629                0                0                  0          0            0              (NULL)
                     15         9923            2         0          0                         1                0            55629                0                0                  0          0            0              (NULL)
```
These are the 16 rows that both your 16 queries and my single query I suggested above will remove. Are you trying to make sure you don't delete non-TC data from custom DBs? 😄

(feel free to argue for keeping your elaborate queries, I am sure there are scenarios I have not thought about yet) :-)

ccrs committed:
object

ccrs commented:
a mix between me being lazy (they come from mysql workbench) and thinking ahead (maybe the script can be done with a couple conditionscripts)

ccrs committed:
hello there

ccrs committed:
build error: missing declaration

ccrs committed:
build error: cannot forward declare a enum

ccrs committed:
build error: unnused variable + missing header

illfated commented:
Happy to see that you found some time to work on it.
Best of luck with solving the build errors.
(Sorry, I don't have any suggestions.)

ccrs committed:
battlefield: add handlekill hook

ccrs committed:
spell_wintergrasp_defender_teleport: revisit conditions

ccrs committed:
battlefield: add virtual mount-related check methods

ccrs committed:
shareddefines: remove type declaration on pvpteamid and teamid

scared of wrong usages on future contributions, so just shield everything (at the cost of one extra header include)

jackpoz commented:
btw was the previous implementation that bad that needed a full rewrite ?

ForesterDev commented:
![123456](https://user-images.githubusercontent.com/11771800/54124800-c1a37980-441c-11e9-9bea-f1300eaa0762.png)
On 3.3.5 players were able to queue in this window, using button near the timer. But I can't find opcode, which activate this posibility for players. Maybe you know how?

ccrs commented:
@jackpoz pretty much yes.
- queue implementation (in, out, group control,..) was all messed up, incomplete and caused several crashes
- spawn system was a joke and relied on SpellArea system which caused absurd ingame behaviour
- battle control through commands was a mere placeholder, no real functionality was actually working
- packet handling was wrong: hacky in some places, reused packets to covert different use cases..

the most decent parts of the previous impl were:
- specific objects such as graveyard and capturepoints (and most of it were copypastes from battleground)
- other minor definitions

eitherway, the real challenge of this PR is to commit the respawn of the area...
I'm leaning towards using SpawnGroups... when the time comes (soon TM) I'll shout out/cry/ for help :smile:

ccrs commented:
@ForesterDev I cant remember which exactly, but my memory is telling me that its the same opcode sent by queue npcs.... though I wouldnt trust my memory capacity :smile:

ccrs committed:
battlefield: implement IsSpellAreaAllowed

ccrs committed:
battlefield: rename mount check methods and add missing impl on wg

ccrs committed:
battlefield: add resurrection handlers

ccrs committed:
player: removepvpinsignia

ccrs committed:
dbcenums: rename AREA_FLAG_WINTERGRASP_2

ccrs committed:
battlefield: getclosestgraveyard

illfated commented:
Just in case you decide to keep your current SQL queries to delete the rows, one by one, from the `conditions` table, I suggest shortening it to contain only the meaningful values:
```sql
DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 14 AND `SourceGroup` = 9904 AND `SourceEntry` = 13759 AND `ElseGroup` = 0 AND `ConditionValue1` = 33280;
DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 14 AND `SourceGroup` = 9904 AND `SourceEntry` = 13759 AND `ElseGroup` = 1 AND `ConditionValue1` = 55629;
DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 14 AND `SourceGroup` = 9904 AND `SourceEntry` = 13761 AND `ElseGroup` = 0 AND `ConditionValue1` = 55629;
DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 14 AND `SourceGroup` = 9904 AND `SourceEntry` = 13761 AND `ElseGroup` = 0 AND `ConditionValue1` = 33280;
DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 14 AND `SourceGroup` = 9923 AND `SourceEntry` = 13798 AND `ElseGroup` = 1 AND `ConditionValue1` = 55629;
DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 14 AND `SourceGroup` = 9923 AND `SourceEntry` = 13798 AND `ElseGroup` = 0 AND `ConditionValue1` = 33280;
DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 14 AND `SourceGroup` = 9923 AND `SourceEntry` = 14172 AND `ElseGroup` = 0 AND `ConditionValue1` = 55629;
DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 14 AND `SourceGroup` = 9923 AND `SourceEntry` = 14172 AND `ElseGroup` = 0 AND `ConditionValue1` = 33280;
DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 15 AND `SourceGroup` = 9904 AND `SourceEntry` =     0 AND `ElseGroup` = 0 AND `ConditionValue1` = 33280;
DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 15 AND `SourceGroup` = 9904 AND `SourceEntry` =     0 AND `ElseGroup` = 1 AND `ConditionValue1` = 55629;
DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 15 AND `SourceGroup` = 9904 AND `SourceEntry` =     1 AND `ElseGroup` = 0 AND `ConditionValue1` = 55629;
DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 15 AND `SourceGroup` = 9904 AND `SourceEntry` =     2 AND `ElseGroup` = 0 AND `ConditionValue1` = 55629;
DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 15 AND `SourceGroup` = 9923 AND `SourceEntry` =     0 AND `ElseGroup` = 0 AND `ConditionValue1` = 33280;
DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 15 AND `SourceGroup` = 9923 AND `SourceEntry` =     0 AND `ElseGroup` = 1 AND `ConditionValue1` = 55629;
DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 15 AND `SourceGroup` = 9923 AND `SourceEntry` =     1 AND `ElseGroup` = 0 AND `ConditionValue1` = 55629;
DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 15 AND `SourceGroup` = 9923 AND `SourceEntry` =     2 AND `ElseGroup` = 0 AND `ConditionValue1` = 55629;
```
I still maintain it is just as valid to use the simplest form to remove that very same content:
```sql
DELETE FROM `conditions` WHERE `SourceGroup` IN (9904, 9923);
```
But it is your PR, so I will wait and see what you decide to do in the end. 👍

ccrs committed:
utilities: countdowntimer

ccrs committed:
player: couple battlefield related-ish corrections

ccrs committed:
battlefieldmgr: foreach implementation & reorder

ccrs committed:
battlefield: implement fillinitialworldstates & sendglobalworldstates

ccrs committed:
remove uneded declaration

ccrs committed:
spell_gen_mount: add battlefield mount status check

ccrs committed:
spell_wintergrasp_waiting_to_resurrect

illfated commented:
Lots of good work. Happy to see the changes. 👍

ccrs committed:
movementhandler: couple corrections regarding far teleport & battlefield

ccrs committed:
csdebug: correct method calls

ccrs committed:
fix build

ccrs committed:
rename + space before comment

ccrs committed:
fix build

ccrs commented:
well... I cant avoid it anymore... the time has come... the time to check all creatures and gameobjects' spawns.. t.t

pls halp
@Killyana

ccrs committed:
battlefieldbuilding placeholder

ccrs committed:
battlefieldmgr missing parameters on battlefieldwintergrasp creation

ccrs committed:
battlefieldentities battlefieldbuilding constructor + corrections

ccrs committed:
Revert "battlefieldmgr missing parameters on battlefieldwintergrasp creation"

This reverts commit 98ac9e57269b0ff15ed1beccdf106d751cb04fae.

ccrs committed:
spellmgr use the most adequate call to battlefieldmgr

illfated commented on sql/updates/world/3.3.5/2019_99_99_00_world.sql:
```diff
@@ -0,0 +1,44 @@
+DELETE FROM `spell_area` WHERE `spell` = 56618 AND `area` = 4538 AND `quest_start` = 0 AND `aura_spell` = 0 AND `racemask` = 0 AND `gender` = 2;
+DELETE FROM `spell_area` WHERE `spell` = 56618 AND `area` = 4539 AND `quest_start` = 0 AND `aura_spell` = 0 AND `racemask` = 0 AND `gender` = 2;
+DELETE FROM `spell_area` WHERE `spell` = 56618 AND `area` = 4611 AND `quest_start` = 0 AND `aura_spell` = 0 AND `racemask` = 0 AND `gender` = 2;
+DELETE FROM `spell_area` WHERE `spell` = 56618 AND `area` = 4612 AND `quest_start` = 0 AND `aura_spell` = 0 AND `racemask` = 0 AND `gender` = 2;
```
Simplify 4 lines by merging them into 1 (doing the exact same thing):
```sql
DELETE FROM `spell_area` WHERE `spell` = 56618 AND `area` IN (4538,4539,4611,4612) AND `quest_start` = 0 AND `aura_spell` = 0 AND `racemask` = 0 AND `gender` = 2;
```

illfated commented on sql/updates/world/3.3.5/2019_99_99_00_world.sql:
```diff
@@ -0,0 +1,44 @@
+DELETE FROM `spell_area` WHERE `spell` = 56617 AND `area` = 4538 AND `quest_start` = 0 AND `aura_spell` = 0 AND `racemask` = 0 AND `gender` = 2;
+DELETE FROM `spell_area` WHERE `spell` = 56617 AND `area` = 4539 AND `quest_start` = 0 AND `aura_spell` = 0 AND `racemask` = 0 AND `gender` = 2;
+DELETE FROM `spell_area` WHERE `spell` = 56617 AND `area` = 4611 AND `quest_start` = 0 AND `aura_spell` = 0 AND `racemask` = 0 AND `gender` = 2;
+DELETE FROM `spell_area` WHERE `spell` = 56617 AND `area` = 4612 AND `quest_start` = 0 AND `aura_spell` = 0 AND `racemask` = 0 AND `gender` = 2;
```
Simplify 4 lines by merging them into 1 (doing the exact same thing):
```sql
DELETE FROM `spell_area` WHERE `spell` = 56617 AND `area` IN (4538,4539,4611,4612) AND `quest_start` = 0 AND `aura_spell` = 0 AND `racemask` = 0 AND `gender` = 2;
```

Kittnz commented:
Will we have database spawns for these creatures and gameobjects instead of the hardcoded ones?

Killyana commented:
Need info about the phasing used when the BG in ON and OFF

illfated commented:
Ah, right, of course. https://wow.gamepedia.com/Wintergrasp#NPCs_and_monsters:
> The side that controls Wintergrasp Keep will be able to see elemental Revenants, with roughly twice as much health as the regular elementals and a guaranteed droprate for crystallized elements.

ccrs commented:
NPC phasing system makes it almost inviable to use here on wintergrasp: factory areas are simply too small, one step aside and everything disappears in front of your character.

The only workaround possible without falling deeply into the hack area, is too use SpawnGroups. Yes, I know, auras should be used, every other solution is not blizzlike... but:
- I will use auras whenever possible
- SpawnGroups are easily adaptable into aura stuffs (once it can be actually bearable)

So yeah, not a single hardcoded spawn, everything must be in the DB :+1:

ccrs commented:
The best way to know what are we dealing with, is just to checkout this PR, enable GM Mode, and move around WG.

illfated commented:
Isn't it possible to set the phasing to cover the entire Wintergrasp zone (including sub-zones) so players don't leave the phase unless they go to a neighboring zone like Icecrown, Dragonblight, Borean Tundra or Sholazar Basin?

ccrs committed:
fix build

Kittnz commented:
fyi, don't forget to implement the world states (ie pvp detection)

ccrs committed:
fix build

ccrs commented:
```
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\include\xmemory0(881): error C2280: 'std::pair<const _Kty,_Ty>::pair(const std::pair<const _Kty,_Ty> &)': attempting to reference a deleted function [C:\projects\trinitycore\build\src\server\game\game.vcxproj]
```
```
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\include\memory(2082): error C2027: use of undefined type 'Battlefield' (compiling source file C:\projects\trinitycore\src\server\game\World\World.cpp) [C:\projects\trinitycore\build\src\server\game\game.vcxproj]
  c:\projects\trinitycore\src\server\game\groups\Group.h(29): note: see declaration of 'Battlefield' (compiling source file C:\projects\trinitycore\src\server\game\World\World.cpp)
```
source: https://github.com/ccrs/TrinityCore/blob/battlefield-1/src/server/game/Battlefield/BattlefieldMgr.h

whats the deal with AppVeyor here? @Shauren @jackpoz

jackpoz commented:
did you try building locally with same AppVeyor cmake settings, whatever they are ? (probably PCH and Dynamic Scripts)

ccrs commented:
nope, I'm actually using circleci nopch as "reference"
kudos to you and them btw, you for setting up the build script and them for that awesome build time

ccrs committed:
fix build + more battlefieldentity placeholders

illfated commented:
Looks like the cleanup is coming along nicely, good job. Sorry that I can't help with the AppVeyor failures, looks too messy for me to understand the errors.

Shauren commented:
You were supposed to **add** a deleted copy constructor to BattlefieldMgr, not remove it from everything else

ccrs committed:
fix build

illfated commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -4990,32 +4990,32 @@ void Player::RepopAtGraveyard()
-        TeleportTo(ClosestGrave->map_id, ClosestGrave->x, ClosestGrave->y, ClosestGrave->z, GetOrientation(), shouldResurrect ? TELE_REVIVE_AT_TELEPORT : 0);
-        if (isDead())                                        // not send if alive, because it used in TeleportTo()
+        TeleportTo(ClosestGraveyard->map_id, ClosestGraveyard->x, ClosestGraveyard->y, ClosestGraveyard->z, GetOrientation(), shouldResurrect ? TELE_REVIVE_AT_TELEPORT : 0);
+        if (isDead()) // not send if alive, because it used in TeleportTo()
```
This may look superficial and cosmetic, but since you already moved the comment, could you add 2 words to make the sentence more complete?
```suggestion
        if (isDead()) // do not send if alive, because it is used in TeleportTo()
```

ccrs replied:
I'm all aboard the grammar correction train 👍

ccrs committed:
fix build

ccrs committed:
battlefieldentities another placeholder

Killyana commented:
Missing guards on wintergrasps https://gist.github.com/Killyana/b2fe4715d8ed3c83ecbf9b5c67fb6168  it contains also spawns that was hard-coded on the cpp, still some phase adjustment mut be one on existing spawns.

Shauren commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -6695,6 +6695,42 @@ void Player::RewardReputation(Quest const* quest)
+ *  If in a battleground or battlefield player dies, and an enemy removes the insignia, the player's bones is lootable
+ *  Called by remove insignia spell effect
+ */
+void Player::RemovePVPInsignia(Player* looter)
```
Is this actually a thing for wintergrasp? Was definitely not true in tol barad and ashran

ccrs replied:
see this "reliable" source https://www.mmo-champion.com/threads/768794-Do-you-remove-insignias
omegalul

ccrs committed:
cleanup sql

ccrs committed:
grammarrrrrrrrrrrrrr

ccrs committed:
battlefieldentities more structure

illfated commented:
Thank you for the SQL cleanup (in addition to your other changes), it looks a lot more straight-forward now. 👍

ccrs committed:
Walls spawns + basic control over them

ccrs committed:
fix build

ccrs committed:
fix build

AByakugans commented:
Is there anything for us to test already or you want to merge it to 335 branch directly?

ccrs commented:
@AByakugans we are now in the "spawn phase": we must spawn everything that was hardcoded or missing, and thats a lot of work.
So any help in that regard (spawning creatures and gameobjects) is VERY much appreciated :heart:

ccrs committed:
battlefieldentity rename and add onremove method

ccrs committed:
battlefieldentities: more definitions

ccrs committed:
zonewintergrasp: change static array to static vector

jackpoz commented:
hey @ccrs could you mark this PR as draft in GitHub ? it makes easier to see which PR are ready :) thanks

ccrs commented:
how would I do that?

illfated commented:
I have only used the Draft feature when opening a new PR, but I don't remember seeing that feature available after setting the PR to "Ready for review". 
(I just checked my PR in a different repository, I can't find any option to do so even if I open anything for editing.)

ccrs committed:
battlefieldentities: more work into those constructors + change EntityEntry into an unordered_set

illfated commented:
Lots of nice changes in here, going to get a test build to run for myself this weekend to see the changes.

ccrs committed:
battlefieldgraveyard: basic support for wintergrasp

ccrs committed:
fix build

ccrs committed:
fix build

ccrs committed:
fix build

ccrs commented:
another tricky one to find 💃

ccrs committed:
battlefieldgraveyard: further support using the on create and remove hooks

ccrs committed:
battlefieldwg: fix typo

ccrs committed:
DB: spawn teleporters

illfated commented:
Happy to see every update and commit from you. Have you got any pointers to how to test your PR content?

I have forked your PR branch and compiled the servers (also updated maps/vmaps after the changes this last weekend), but when it comes to looking for stuff in Wintergrasp, I find it difficult to know what to look for (I never actually participated in WG on retail). I only dropped by a few times to farm revenants for Eternals (and the crystallized bits). Is the PR content still considered to be too "bare-bone" to actually test for anything, or is it useful to report that my character got dismounted over WG when there was no battle going on? (Not sure if this is intended & expected behaviour, because my retail experience from WG is close to non-existing.)

ccrs commented:
Was about to answer that no in-game testing is actually necessary at this point, but I stand corrected. What you experienced about mounts is not intended, thanks for reporting 👍 

We're still on same stage as couple weeks ago, spawning. Need to DB spawn all the missing creatures and gameobjects, and use spawn groups whenever necessary.

ccrs committed:
wintergrasp: revisit flying restriction aura conditions

ccrs committed:
wintergrasp: add worldsafelocs entries

AByakugans commented:
Hello @ccrs is there anything new? I dont know how to get the sql you need or you can tell me how maybe if I can be of help here... This is great part if the game it would be shame to not have it working... Already more than half year here dusting...

illfated commented:
Thank you for keeping the interest alive regarding Wintergrasp, @AByakugans . You do not necessarily have to provide any SQL content, but it would be nice if you could make some sort of lists of static,  roaming NPCs and if there is useful spawn information either in web pages or verified retail videos. I guess there will be more feedback sooner or later regarding what we should focus on doing (and maybe in what order as well).

ccrs commented:
Dont worry, I did not forget 😃

illfated commented:
Good weekend to you, @ccrs . Would you mind resolving the file conflict?

Killyana commented:
I found this 2 spells, but not sure how it's used:
``` sql
https://www.wowhead.com/spell=58554/lok-regar
https://www.wowhead.com/spell=58555/great-honor

AByakugans commented:
Hello, anything new? Version for us to test? @ccrs

illfated commented:
First, we need to resolve the file conflict to get the PR content usable on current 3.3.5 branch to make it easy to fork for testing. (if you are ready to test, you could choose to fix the conflicts yourself when forking this branch (https://github.com/ccrs/TrinityCore/tree/battlefield-1).

Secondly, notice that ccrs mentioned in https://github.com/TrinityCore/TrinityCore/pull/22700#issuecomment-432274468 "DB spawning will take some time, [ ... ]", indicating that we still need verified DB content for the actual Wintergrasp spawns, because the current ones could be different from the retail spawns (number of spawns, react state, coordinates, orientation, pathing).

Drahy commented:
Can @Killyana maybe help here with the spawns?

Killyana commented:
I did: https://gist.github.com/Killyana/b2fe4715d8ed3c83ecbf9b5c67fb6168
But I don't have a retail account, so I cannot check if some guards spawns only when the battle starts or not.

jackpoz commented:
Closing this PR as it seems abandoned. Feel free to ask here to reopen it if anyone would like to work on it.

offl commented:
Closed-Abandoned label?
